src/spatial/2d.js

```diff
@@ -1585,25 +1606,41 @@ Crafty.c("Motion", {
     _linearMotionTick: function(frameData) {
         var dt = frameData.dt/1000;
 
-        var oldX = this._x;
-        var oldY = this._y;
+        var oldDirection = this.__oldDirection;
+        var _vx = this._vx, dvx = _vx >> 31 | -_vx >>> 31, // Math.sign(this._vx)
+            _vy = this._vy, dvy = _vy >> 31 | -_vy >>> 31; // Math.sign(this._vy)
+        if (oldDirection.x !== dvx || oldDirection.y !== dvy) {
+            var directionEvent = this.__directionEvent;
+            directionEvent.x = oldDirection.x = dvx;
+            directionEvent.y = oldDirection.y = dvy;
+            this.trigger('NewDirection', directionEvent);
+        }
+
+        var oldX = this._x, vx = this._vx, ax = this._ax,
+            oldY = this._y, vy = this._vy, ay = this._ay;
+
         // s += v * Δt + (0.5 * a) * Δt * Δt
-        var newX = oldX + this._vx * dt + 0.5 * this._ax * dt * dt;
-        var newY = oldY + this._vy * dt + 0.5 * this._ay * dt * dt;
+        var newX = oldX + vx * dt + 0.5 * ax * dt * dt;
+        var newY = oldY + vy * dt + 0.5 * ay * dt * dt;
         // v += a * Δt
-        this.vx = this._vx + this._ax * dt;
-        this.vy = this._vy + this._ay * dt;
+        this.vx = vx + ax * dt;
+        this.vy = vy + ay * dt;
         // Δs = s[t] - s[t-1]
         this._dx = newX - oldX;
         this._dy = newY - oldY;
 
+        var movedEvent = this.__movedEvent;
         if (this._dx !== 0) {
             this.x = newX;
-            this.trigger('Moved', {x: oldX, y: newY});
+            movedEvent.x = oldX;
+            movedEvent.y = newY;
+            this.trigger('Moved', movedEvent);
```

Comment:
I don't *think* this will matter, but the way we've recommended handling collisions is to listen to the "Moved" event, test for collisions, and then move back to the old position if necessary.

The catch here is that this will cause a second "Moved" event to trigger on the entity before the first has returned, and both of them use the same `_movedEvent` object.  As long as the collision check is the *only* thing listening to Moved, no big deal, but it might cause some unexpected behavior if someone has multiple event handlers for "Moved".

I think this would be pretty rare, so I'm oksolving this particular issue down the road.  But maybe add a comment about it?